### DISCUSSION TITLE
eneble cors with trpc and next

### DISCUSSION THREAD

**Discussion by ifeanyidotdev** [trpc/trpc#2019] [Q&A]
i have a site in next hosted on netlify but i can't make any request because the cors blocking, how  can i solve this problem

**Comment by KATT** [comment]
Paste the question you wrote above on google and that should provide better answer than what we can do here

**Reply by KATT** [reply]
I think you need to wrap the `[trpc].tsx`-api handler in some CORS thing

https://nextjs.org/docs/api-routes/api-middlewares#connectexpress-middleware-support

**Reply by joecoker** [reply]
@Mremmalex Have you managed to get this working? I've run into the same issue hosting on Amplify. I'm unsure how to add cors to the handler returned by createNextApiHandler.

**Reply by KATT** [reply]
Never used amplify. Maybe the Lambda-adapter can help you https://github.com/trpc/trpc/tree/main/packages/server/adapters/lambda

I don't have the bandwidth to provide more guidance unfortunately

**Reply by ifeanyidotdev** [reply]
@joecoker go to your _app.js change the remote your to the amplify url , thats was my problem i was targeting a url that was not set

**Reply by wiredmatt** [reply]
I can't get cors working on local development neither, I tried the following

```tsx
// src/pages/api/trpc/[trpc].ts
import { createNextApiHandler } from '@trpc/server/adapters/next'
import { appRouter } from '../../../server/router'
import { createContext } from '../../../server/router/context'

// export API handler
export default createNextApiHandler({
  router: appRouter,
  createContext: createContext,
  responseMeta(ctx) {
    ctx.ctx?.res?.setHeader(
      'Access-Control-Allow-Origin',
      '.brand.localhost'
    )
    ctx.ctx?.res?.setHeader('Access-Control-Request-Method', '*')
    ctx.ctx?.res?.setHeader(
      'Access-Control-Allow-Methods',
      'OPTIONS, GET'
    )
    ctx.ctx?.res?.setHeader('Access-Control-Allow-Headers', '*')

    if (ctx.ctx?.req?.method === 'OPTIONS') {
      ctx.ctx?.res?.writeHead(200)
    }

    return {
      headers: ctx.ctx?.res?.getHeaders() as Record<
        string,
        string
      >,
      statusCode: ctx.ctx?.res?.statusCode || 200
    }
  }
})
```

I then have another next.js app from which I would like to consume this TRPC API, is it possible?

**Reply by joaobibiano** [reply]
> 

It didn't work here 😢 

Have you managed another way to make it works?

**Reply by wiredmatt** [reply]
Nope, I gave up on trpc as I doesn't suit basic business requirements for now, or at least they are not well documented. I'm using a custom hooks generator from an open-api schema, [this](https://blog.logrocket.com/generating-integrating-openapi-services-react/) was my starting point. I have recently stumbled across [this](https://github.com/ecyrbe/zodios) package, which does essentially the same trpc does but in a more traditional way. Finally there's [this other](https://github.com/RobinTail/express-zod-api) which is intended to work with an expressjs backend. I think both of these options are better for scaling and if you have multiple microfrontends, I couldn't get trpc to work due to the cors issue

**Comment by fortezhuo** [comment]
Hi, 

Currently, I am playing tRPC and NextJS. And here the wrapper.

```
import type { NextApiHandler, NextApiRequest, NextApiResponse } from 'next';
import Cors from 'cors';

const cors = Cors();

function runMiddleware(req: NextApiRequest, res: NextApiResponse, fn: any) {
  return new Promise((resolve, reject) => {
    fn(req, res, (result: any) => {
      if (result instanceof Error) {
        return reject(result);
      }

      return resolve(result);
    });
  });
}

export function withCors(handler: NextApiHandler) {
  return async (req: NextApiRequest, res: NextApiResponse) => {
    await runMiddleware(req, res, cors);

    return await handler(req, res);
  };
}
```

And [trpc].ts

```
export default withCors(
  trpcNext.createNextApiHandler({
    router: appRouter,
    ....
    ....
}))
```

Hope this snippet solves your problem

**Reply by Jaaneek** [reply]
@fortezhuo Is there anything else needed to make it work?  It works on localhost but does not on vercel deployment

**Reply by fortezhuo** [reply]
@Jaaneek 
You can take a look at https://vercel.com/guides/how-to-enable-cors . Hope the link solves your problem

**Comment by shahzaibfs** [comment]
You can use cors like that. 
haven't tested this personlly but you can assign response headers like that . 
and please do let me know if you found this helpful. 

![image](https://github.com/trpc/trpc/assets/84914596/89a57a6c-b944-469d-b90d-558d49bc57d8)

**Reply by kippytech** [reply]
@shahzaibfs where did you get or import from that createTRPCContext? I'm getting an error as I try to follow your code snippet

**Comment by zendostrike** [comment]
For Nest.js: Be sure to enable cors before TRCP applyMiddleware. If you set enable cors after trcp, your trcp requests won't work.

```
const app = await NestFactory.create(AppModule);
app.enableCors();

const trpc = app.get(TrpcRouter);
trpc.applyMiddleware(app);
```
